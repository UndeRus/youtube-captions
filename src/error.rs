use thiserror::Error;

/// A `Result` alias where the `Err` case is `youtube_captions::Error`.
pub type Result<T, E = Error> = std::result::Result<T, E>;

/// The Errors that may occur while scraping captions.
#[derive(Debug, Error)]
pub enum Error {
  #[error("the video is no longer available")]
  VideoUnavailable,

  /// YouTube is receiving too many requests from this IP and now requires solving a captcha to continue.
  /// One of the following things can be done to work around this:
  /// - Manually solve the captcha in a browser and export the cookie
  /// - Use a different IP address
  /// - Wait until the ban on your IP has been lifted
  #[error(
    "YouTube is receiving too many requests from this IP and now requires solving a captcha to continue.
  One of the following things can be done to work around this:
  - Manually solve the captcha in a browser and export the cookie
  - Use a different IP address
  - Wait until the ban on your IP has been lifted"
  )]
  CaptchaRequired,

  /// Subtitles are disabled for this video.
  #[error("Subtitles are disabled for this video")]
  TranscriptsDisabled,

  // /// No transcripts are available for this video.
  // #[error("No transcripts are available for this video")]
  // NoTranscriptAvailable,

  /// The requested file is not translatable.
  #[error("The requested file is not translatable")]
  NotTranslatable,

  /// The requested translation language is not available.
  #[error("The requested translation language is not available")]
  TranslationLanguageNotAvailable,

  /// The cookies provided are not valid (may have expired).
  #[error("The cookies provided are not valid (may have expired)")]
  CookiesInvalid,

  /// Failed to automatically give consent to saving cookies.
  #[error("Failed to automatically give consent to saving cookies")]
  FailedToCreateConsentCookie,

  #[error("Request to YouTube failed: {0}")]
  NetworkError(#[from] reqwest::Error),

  #[error("Invalid JSON: {0}")]
  InvalidJson(#[from] serde_json::Error),

  #[cfg(feature = "quick-xml")]
  #[error("Invalid XML: {0}")]
  InvalidXml(#[from] quick_xml::de::DeError),
}
