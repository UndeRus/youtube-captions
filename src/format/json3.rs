// json3 is pre-parsed and uses the two letter attribute name + full name, so fc -> fcForeColor
// YouTube's captions.js also has both a WebVTT and json3 parser, so it's easy to see what everything means.
// https://github.com/yingted/ytcc2/blob/ece2ae8006274d7cd4aed5ac3b694f778cc9ad10/captions/src/json3.ml#L18

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Transcript;
