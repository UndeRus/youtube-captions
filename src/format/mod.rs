#[cfg(feature = "json3")]
pub mod json3;
#[cfg(feature = "srv1")]
pub mod srv1;
#[cfg(feature = "srv2")]
pub mod srv2;
#[cfg(feature = "srv3")]
pub mod srv3;
#[cfg(feature = "ttml")]
pub mod ttml;

#[derive(Debug, Default, Clone, Copy)]
pub enum Format {
  /// Web Video Text Tracks
  VTT,

  /// Timed Text Markup Language
  TTML,

  /// YouTube Timed Text, v1
  #[default]
  SRV1,

  /// YouTube Timed Text, v2
  SRV2,

  /// YouTube Timed Text, v3
  SRV3,

  JSON3,
}

impl From<Format> for &'static str {
  fn from(value: Format) -> Self {
    match value {
      Format::VTT => "vtt",
      Format::TTML => "ttml",
      Format::SRV1 => "srv1",
      Format::SRV2 => "srv2",
      Format::SRV3 => "srv3",
      Format::JSON3 => "json3",
    }
  }
}
