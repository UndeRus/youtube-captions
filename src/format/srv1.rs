use std::borrow::Cow;
use std::str::FromStr;

use quick_xml::escape::unescape as unescape_xml;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Transcript {
  #[serde(rename(deserialize = "text"))]
  pub segments: Vec<TextSegment>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TextSegment {
  #[serde(rename(deserialize = "@start"))]
  pub start_secs: f32,
  #[serde(rename(deserialize = "@dur"))]
  pub duration_secs: f32,
  #[serde(rename(deserialize = "$value"))]
  pub value: String,
}

impl TextSegment {
  pub fn end_secs(&self) -> f32 {
    self.start_secs + self.duration_secs
  }
}

impl IntoIterator for Transcript {
  type Item = TextSegment;
  type IntoIter = std::vec::IntoIter<Self::Item>;

  fn into_iter(self) -> Self::IntoIter {
    self.segments.into_iter()
  }
}

impl FromStr for Transcript {
  type Err = quick_xml::DeError;

  fn from_str(input: &str) -> Result<Self, Self::Err> {
    #[cfg(not(test))]
    let mut transcript: Transcript = quick_xml::de::from_str(input)?;

    #[cfg(test)]
    let mut transcript: Transcript = {
      let de = &mut quick_xml::de::Deserializer::from_str(input);
      serde_path_to_error::deserialize(de).unwrap()
    };

    for it in &mut transcript.segments {
      if let unescaped @ Cow::Owned(_) = unescape_xml(&it.value).unwrap() {
        it.value = unescaped.into_owned();
      };
    }

    Ok(transcript)
  }
}
