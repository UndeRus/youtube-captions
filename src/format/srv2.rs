use std::borrow::Cow;
use std::str::FromStr;

use quick_xml::escape::unescape as unescape_xml;
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct Transcript {
  #[serde(rename = "$value")]
  pub elements: Vec<Element>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Element {
  Text(TextSegment),
  Window(Window),
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct TextSegment {
  #[serde(rename(deserialize = "@t"))]
  pub timestamp_millis: u32,
  #[serde(default, rename(deserialize = "@d"))]
  pub duration_millis: u32,
  #[serde(default, rename(deserialize = "@append"))]
  pub append: bool,
  #[deprecated(note = "if you know what this field is about, please, raise an issue")] // todo: paste a link to github
  #[serde(default, rename(deserialize = "@r"))]
  pub r: u32,
  #[deprecated(note = "if you know what this field is about, please, raise an issue")] // todo: paste a link to github
  #[serde(default, rename(deserialize = "@c"))]
  pub c: u32,
  #[serde(default, rename(deserialize = "$value"))]
  pub value: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
pub struct Window {
  #[serde(rename(deserialize = "@id"))]
  pub id: u32,
  // todo: may have values "define", ...?
  #[serde(rename(deserialize = "@op"))]
  pub operation: String,
  #[serde(rename(deserialize = "@t"))]
  pub timestamp_millis: u32,
  #[serde(rename(deserialize = "@ap"))]
  pub anchor_point: AnchorPoint,
  /// X from left
  #[serde(rename(deserialize = "@ah"))]
  pub horizontal_alignment: u32,
  /// Y from top
  #[serde(rename(deserialize = "@av"))]
  pub vertical_alignment: u32,
  #[serde(rename(deserialize = "@rc"))]
  pub rows_total: u8,
  /// Each column has en-dash width
  #[serde(rename(deserialize = "@cc"))]
  pub columns_total: u8,
  #[serde(rename(deserialize = "@sd"))]
  pub scroll_direction: ScrollDirection,
  #[serde(default, rename(deserialize = "@pd"))]
  pub print_direction: PrintDirection,
  #[serde(rename(deserialize = "@ju"))]
  pub text_alignment: TextAlignment,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize_repr, Deserialize_repr)]
#[repr(u8)]
pub enum AnchorPoint {
  TopLeft = 0,
  TopCenter = 1,
  TopRight = 2,
  CenterLeft = 3,
  TrueCenter = 4,
  CenterRight = 5,
  BottomLeft = 6,
  BottomCenter = 7,
  BottomRight = 8,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize_repr, Deserialize_repr)]
#[repr(u8)]
pub enum TextAlignment {
  /// Equals to _left_ in LTR
  Start = 0,
  /// Equals to _right_ in LTR
  Eng = 1,
  Center = 2,
  Justify = 3,
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize_repr, Deserialize_repr)]
#[repr(u8)]
pub enum PrintDirection {
  #[default]
  LtrHorizontal = 0,
  RtlHorizontal = 1,
  /// Upright text: lays out the glyphs for vertical scripts naturally (upright),
  /// as well as the characters of horizontal scripts
  VerticalLtr = 2,
  /// Sideways text: causes characters to be laid out as they would be horizontally,
  /// but with the whole line rotated 90Â° clockwise.
  VerticalRtl = 3,
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize_repr, Deserialize_repr)]
#[repr(u8)]
pub enum ScrollDirection {
  #[default]
  LTR = 0,
  RTL = 1,
}

impl FromStr for Transcript {
  type Err = quick_xml::DeError;

  fn from_str(input: &str) -> Result<Self, Self::Err> {
    #[cfg(not(test))]
    let mut transcript: Transcript = quick_xml::de::from_str(input)?;

    #[cfg(test)]
    let mut transcript: Transcript = {
      let de = &mut quick_xml::de::Deserializer::from_str(input);
      serde_path_to_error::deserialize(de).unwrap()
    };

    for it in &mut transcript.elements {
      if let Element::Text(it) = it {
        if let unescaped @ Cow::Owned(_) = unescape_xml(&it.value).unwrap() {
          it.value = unescaped.into_owned();
        };
      }
    }

    Ok(transcript)
  }
}
