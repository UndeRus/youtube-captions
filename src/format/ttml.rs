// https://github.com/arcusmaximus/YTSubConverter/tree/master/YTSubConverter.Shared/Formats/Ttml

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Transcript;
