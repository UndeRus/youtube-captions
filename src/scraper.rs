use std::collections::HashSet;

use language_tags::LanguageTag;
use regex::Regex;
use serde::Deserialize;

use crate::error::{Error, Result};
use crate::format::*;

pub struct DigestScraper {
  cookie: tokio::sync::RwLock<Option<String>>,
  http: reqwest::Client,
}

#[derive(Debug)]
pub struct Digest {
  pub captions: Vec<CaptionScraper>,
  pub can_be_translated_to: HashSet<String>,
}

#[derive(Debug, Clone)]
pub struct CaptionScraper {
  pub(crate) url: String,
  pub(crate) http: reqwest::Client,
  /// `true`, if generated using automatic speech recognition
  pub is_generated: bool,
  /// `true`, if can be translated with [`CaptionScraper::translate_to`] method.
  pub is_translatable: bool,
  pub lang_name: String,
  pub lang_tag: LanguageTag,
}

impl DigestScraper {
  pub fn new(http: reqwest::Client) -> Self {
    Self { http, cookie: tokio::sync::RwLock::new(None) }
  }

  async fn get(&self, url: &str) -> Result<String> {
    let cookie = self.cookie.read().await;
    let cookie = cookie.as_ref().map_or("", std::ops::Deref::deref);
    let res = self.http.get(url).header("Cookie", cookie).send().await?;
    Ok(res.error_for_status()?.text().await?)
  }

  async fn fetch_video_page(&self, video_id: &str, lang: &str) -> Result<String> {
    let url = format!(r#"https://youtube.com/watch?hl={}&persist_hl=1&v={}"#, lang, video_id);
    let mut html = self.get(&url).await?;

    let consent = r#"action="https://consent.youtube.com/s""#;
    if html.contains(consent) {
      let mut cookie = self.cookie.write().await;
      *cookie = Some(Self::extract_consent_cookie(&html)?);

      html = self.get(&url).await?;
      if html.contains(consent) {
        return Err(Error::FailedToCreateConsentCookie);
      }
    }

    Ok(html)
  }

  fn extract_consent_cookie(html: &str) -> Result<String, Error> {
    lazy_static::lazy_static! {
      static ref RE: Regex = Regex::new(r#"name="v" value="(.*?)"#).unwrap();
    }
    RE.captures(html)
      .ok_or(Error::FailedToCreateConsentCookie)
      .map(|caps| format!("CONSENT=YES+{};Domain=.youtube.com", &caps[1]))
  }

  fn extract_captions_json(html: &str) -> Result<RawDigest> {
    let (_, html) = html.split_once(r#""captions":"#).ok_or_else(|| match html {
      html if html.contains(r#"class="g-recaptcha""#) => Error::CaptchaRequired,
      html if !html.contains(r#""playabilityStatus":"#) => Error::VideoUnavailable,
      _ => Error::TranscriptsDisabled,
    })?;

    let (json, _) = html.split_once(r#","videoDetails"#).ok_or(Error::TranscriptsDisabled)?;
    let data: RawData = serde_json::from_str(json)?;
    Ok(data.captions)
  }

  pub async fn fetch<'a, Str: Into<Option<&'a str>>>(&self, video_id: &str, lang: Str) -> Result<Digest> {
    let lang = lang.into().unwrap_or("en");
    let html = self.fetch_video_page(video_id, lang).await?;
    let digest = DigestScraper::extract_captions_json(&html)?;

    let convert = |it: RawCaptionTrack| {
      CaptionScraper {
        url: it.base_url,
        http: self.http.clone(),
        lang_name: it.name.text,
        is_generated: matches!(it.kind.as_deref(), Some("asr")),
        is_translatable: it.is_translatable,
        lang_tag: it.language_code.parse().unwrap(), // we trust Google, don't we?
      }
    };

    Ok(Digest {
      captions: digest.caption_tracks.into_iter().map(convert).collect(),
      can_be_translated_to: digest.translation_languages.into_iter().map(|it| it.code).collect(),
    })
  }
}

impl CaptionScraper {
  /// The parameter value is an [ISO 639-1 two-letter language code] that identifies the desired caption language.
  /// The translation is generated by using machine translation, such as Google Translate.
  ///
  /// [ISO 639-1 two-letter language code]: http://www.loc.gov/standards/iso639-2/php/code_list.php
  pub fn translate_to(&mut self, language: &LanguageTag) -> Result<&mut Self> {
    if self.is_translatable {
      self.url.push_str("&tlang=");
      self.url.push_str(language.as_str());
      Ok(self)
    } else {
      Err(Error::NotTranslatable)
    }
  }

  /// The parameter specifies that the caption track should be returned in a specific format.
  /// If the parameter is not included in the request, the track is returned in its original format.
  pub async fn fetch(&self, format: Format) -> Result<String> {
    let format: &str = format.into();
    let url = format!("{}&fmt={}", &self.url, format);
    Ok(self.http.get(&url).send().await?.text().await?)
  }

  #[cfg(feature = "json3")]
  pub async fn fetch_json3(&self) -> Result<json3::Transcript> {
    unimplemented!()
  }

  #[cfg(feature = "srv1")]
  pub async fn fetch_srv1(&self) -> Result<srv1::Transcript> {
    Ok(self.fetch(Format::SRV1).await?.parse()?)
  }

  #[cfg(feature = "srv2")]
  pub async fn fetch_srv2(&self) -> Result<srv2::Transcript> {
    Ok(self.fetch(Format::SRV2).await?.parse()?)
  }

  #[cfg(feature = "srv3")]
  pub async fn fetch_srv3(&self) -> Result<srv3::Transcript> {
    Ok(self.fetch(Format::SRV3).await?.parse()?)
  }

  #[cfg(feature = "ttml")]
  pub async fn fetch_ttml(&self) -> Result<ttml::Transcript> {
    unimplemented!()
  }
}

#[derive(Debug, Deserialize)]
struct RawData {
  #[serde(rename = "playerCaptionsTracklistRenderer")]
  pub captions: RawDigest, // todo: what if the root element is absent?
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RawDigest {
  pub caption_tracks: Vec<RawCaptionTrack>,
  pub translation_languages: Vec<RawLanguage>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RawCaptionTrack {
  pub base_url: String,
  pub language_code: String,
  #[serde(default)]
  pub is_translatable: bool,
  #[serde(default)]
  pub kind: Option<String>,
  pub name: RawName,
}

#[derive(Debug, Deserialize)]
struct RawLanguage {
  #[serde(rename = "languageCode")]
  pub code: String,
  // #[serde(rename = "languageName")]
  // pub name: RawName,
}

#[derive(Debug, Deserialize)]
struct RawName {
  #[serde(rename = "simpleText")]
  pub text: String,
}
