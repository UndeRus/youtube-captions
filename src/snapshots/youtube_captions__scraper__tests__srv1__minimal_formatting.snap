---
source: src/scraper.rs
expression: transcript
---
{
  "segments": [
    {
      "start_secs": 4.333,
      "duration_secs": 2.1,
      "value": "(\"Just A Girl\" - No Doubt)"
    },
    {
      "start_secs": 8.6,
      "duration_secs": 3.9,
      "value": "♪ Take this pink ribbon\noff my eyes ♪"
    },
    {
      "start_secs": 12.533,
      "duration_secs": 4.567,
      "value": "♪ I'm exposed and\nit's no big surprise ♪"
    },
    {
      "start_secs": 17.133,
      "duration_secs": 4.233,
      "value": "♪ Don't you think I know\nexactly where I stand? ♪"
    },
    {
      "start_secs": 21.4,
      "duration_secs": 4.766,
      "value": "♪ This world is forcing me\nto hold your hand ♪"
    },
    {
      "start_secs": 26.2,
      "duration_secs": 4.1,
      "value": "♪ 'Cause I'm just a girl,\noh, little old me ♪"
    },
    {
      "start_secs": 30.333,
      "duration_secs": 4.3,
      "value": "♪ Well, don't let me\nout of your sight ♪"
    },
    {
      "start_secs": 34.666,
      "duration_secs": 4.634,
      "value": "♪ Oh, I'm just a girl,\nall pretty and petite ♪"
    },
    {
      "start_secs": 39.333,
      "duration_secs": 4.967,
      "value": "♪ So don't let me\nhave any rights ♪"
    },
    {
      "start_secs": 44.333,
      "duration_secs": 5.0,
      "value": "♪ Oh, I've had it up to here ♪"
    },
    {
      "start_secs": 53.266,
      "duration_secs": 3.334,
      "value": "♪ The moment that\nI step outside ♪"
    },
    {
      "start_secs": 56.633,
      "duration_secs": 4.433,
      "value": "♪ So many reasons\nfor me to run and hide ♪"
    },
    {
      "start_secs": 61.1,
      "duration_secs": 4.3,
      "value": "♪ I can't do the little\nthings I hold so dear ♪"
    },
    {
      "start_secs": 65.433,
      "duration_secs": 5.133,
      "value": "♪ 'Cause it's all those\nlittle things that I fear♪"
    },
    {
      "start_secs": 70.6,
      "duration_secs": 4.1,
      "value": "♪ 'Cause I'm just a girl,\nI'd rather not be ♪"
    },
    {
      "start_secs": 74.733,
      "duration_secs": 4.667,
      "value": "♪ 'Cause they won't\nLet me drive late at night ♪"
    },
    {
      "start_secs": 79.433,
      "duration_secs": 4.067,
      "value": "♪ Oh, I'm just a girl,\nguess I'm some kind of freak ♪"
    },
    {
      "start_secs": 83.533,
      "duration_secs": 4.9,
      "value": "♪ 'Cause they all sit and\nstare with their eyes ♪"
    },
    {
      "start_secs": 88.466,
      "duration_secs": 1.834,
      "value": "♪ Oh, I'm just a girl ♪"
    },
    {
      "start_secs": 90.333,
      "duration_secs": 2.167,
      "value": "♪ Take a good look at me ♪"
    },
    {
      "start_secs": 92.533,
      "duration_secs": 4.8,
      "value": "♪ Just your typical prototype ♪"
    },
    {
      "start_secs": 97.366,
      "duration_secs": 5.9,
      "value": "♪ Oh, I've had it up to here ♪"
    },
    {
      "start_secs": 119.666,
      "duration_secs": 5.667,
      "value": "Oh, am I making myself clear? ♪"
    },
    {
      "start_secs": 128.666,
      "duration_secs": 1.9,
      "value": "♪ I'm just a girl ♪"
    },
    {
      "start_secs": 133.3,
      "duration_secs": 3.9,
      "value": "♪ I'm just a girl in the world ♪"
    },
    {
      "start_secs": 137.233,
      "duration_secs": 4.4,
      "value": "♪ That's all that\nyou'll let me be ♪"
    },
    {
      "start_secs": 141.666,
      "duration_secs": 3.934,
      "value": "♪ Oh, I'm just girl,\nliving in captivity ♪"
    },
    {
      "start_secs": 145.633,
      "duration_secs": 4.8,
      "value": "♪ Your rule of thumb\nmakes me worrisome ♪"
    },
    {
      "start_secs": 150.466,
      "duration_secs": 4.034,
      "value": "♪ Oh, I'm just a girl,\nwhat's my destiny? ♪"
    },
    {
      "start_secs": 154.533,
      "duration_secs": 5.0,
      "value": "♪ What I've succumbed to\nis making me numb ♪"
    },
    {
      "start_secs": 159.566,
      "duration_secs": 3.834,
      "value": "♪ Oh, I'm just a girl,\nmy apologies ♪"
    },
    {
      "start_secs": 163.433,
      "duration_secs": 4.9,
      "value": "♪ What I've become\nis so burdensome ♪"
    },
    {
      "start_secs": 168.366,
      "duration_secs": 4.234,
      "value": "Oh, I'm just a girl, lucky me ♪"
    },
    {
      "start_secs": 172.633,
      "duration_secs": 4.9,
      "value": "♪ Twiddle-dum,\nthere's no comparison ♪"
    },
    {
      "start_secs": 177.566,
      "duration_secs": 4.634,
      "value": "♪ Oh, I've had it up to ♪"
    },
    {
      "start_secs": 182.233,
      "duration_secs": 4.267,
      "value": "♪ Oh, I've had it up to ♪"
    },
    {
      "start_secs": 186.533,
      "duration_secs": 2.733,
      "value": "♪ Oh ♪"
    },
    {
      "start_secs": 189.3,
      "duration_secs": 5.466,
      "value": "♪ I've had it up to here ♪"
    }
  ]
}
